### PR TITLE
Asrc 72 - Add retrieval for Potassium, HgA1C, and Glucose.

### PR DESCRIPTION
**Pull request by jamesbigmac**

### CODE REVIEW

File: srcalc/src/main/java/gov/va/med/srcalc/domain/calculation/ValueRetriever.java [hunk 1]
```diff
@@ -224,6 +224,33 @@ public String retrieveReferenceInfo(final Patient patient, final Variable var)
...
             return variableRefInfo.toString();
         }
+    },
+    POTASSIUM
+    {
+        @Override
+        public void execute(final Patient patient, final VariableEntry variableEntry, final Variable variable,
+            final String key)
+        {
+            addLabValue("POTASSIUM", patient, variableEntry, key);
```
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/domain/calculation/ValueRetriever.java]: You know, there is a better solution than just including these literal strings here.

Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/domain/calculation/ValueRetriever.java]: The unit tests only run `ValueRetriever.WBC` right now. I think we should update `testWithRetrievedLabs()` to test "retrieving" all supported labs.

Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/domain/calculation/ValueRetriever.java]: Don't you think this is a bit excessive to test all of them? We are up to 12 labs.

File: srcalc/src/main/java/gov/va/med/srcalc/domain/VistaLabs.java [hunk 2]
```diff
@@ -111,12 +111,36 @@
...
+    {
+        @Override
+        public List<String> getPossibleLabNames()
+        {
+            return ImmutableList.of("GLUCOSE", "SERUM GLUCOSE");
+        }
     };
     
-    abstract List<String> getPossibleLabNames();
+    public abstract List<String> getPossibleLabNames();
```
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/domain/VistaLabs.java]: This method will need a Javadoc: might as well add it now.

File: srcalc/src/main/java/gov/va/med/srcalc/domain/calculation/ValueRetriever.java [hunk 3]
```diff
@@ -108,7 +109,7 @@ public void execute(final Patient patient, final VariableEntry variableEntry, fi
         public void execute(final Patient patient, final VariableEntry variableEntry, final Variable variable,
                 final String key)
         {
-            addLabValue("ALBUMIN", patient, variableEntry, key);
+            addLabValue(VistaLabs.ALBUMIN, patient, variableEntry, key);
```
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/domain/calculation/ValueRetriever.java]: Ah, this is much better.

File: srcalc/src/test/java/gov/va/med/srcalc/web/view/VariableEntryTest.java [hunk 4]
```diff
@@ -16,6 +19,30 @@
  */
 public class VariableEntryTest
 {
+    /**
+     * A map where the key is the variable key and the value is the ValueRetriever belonging
+     * to that variable.
+     */
+    private static final Map<String, ValueRetriever> LAB_MAP;
+    static {
```
Comment (dgtombs) [inline srcalc/src/test/java/gov/va/med/srcalc/web/view/VariableEntryTest.java]: If you used `ImmutableMap`, you wouldn't need this static initializer block. They are a little obscure and best avoided.

File: srcalc/src/test/java/gov/va/med/srcalc/web/view/VariableEntryTest.java [hunk 5]
```diff
@@ -75,28 +102,52 @@ public final void testRetrievedCardiacAge()
...
-        expected.put(
-                VariableEntry.makeRetrievalString(
-                        VariableEntry.makeNumericalInputName(vars.get(5).getKey())),
-                retrievalString);
+        addAllLabs(expected, vars, patient);
+        final VariableEntry entry = VariableEntry.withRetrievedValues(vars, patient);
         assertEquals(expected, entry.getDynamicValues());
     }
+    
+    private void addAllLabs(final Map<String, String> expected, final List<AbstractVariable> vars,
```
Comment (dgtombs) [inline srcalc/src/test/java/gov/va/med/srcalc/web/view/VariableEntryTest.java]: This method is complex enough to warrant a Javadoc. I had to read it through to understand what it was doing.

It's pretty specific to `testWithRetrievedLabs()`, does it really deserve to be a separate method at all?

File: srcalc/src/test/java/gov/va/med/srcalc/web/view/VariableEntryTest.java [hunk 6]
```diff
@@ -75,28 +102,52 @@ public final void testRetrievedCardiacAge()
...
+            expected.put(VariableEntry.makeNumericalInputName(key),
+                    String.valueOf(labValue.getValue()));
+            expected.put(
+                    VariableEntry.makeRetrievalString(
+                            VariableEntry.makeNumericalInputName(key)),
+                    retrievalString);
+        }
+    }
+    
+    private DiscreteNumericalVariable makeLabVariable(final String key, final ValueRetriever retriever)
```
Comment (dgtombs) [inline srcalc/src/test/java/gov/va/med/srcalc/web/view/VariableEntryTest.java]: This method could use a Javadoc as well.

File: srcalc/src/test/java/gov/va/med/srcalc/web/view/VariableEntryTest.java [hunk 7]
```diff
@@ -75,28 +102,52 @@ public final void testRetrievedCardiacAge()
     @Test
     public final void testWithRetrievedLabs()
```
Comment (dgtombs) [inline srcalc/src/test/java/gov/va/med/srcalc/web/view/VariableEntryTest.java]: I really like this test now: it tests quite a bit of functionality with a good economy of code.

File: srcalc/src/test/java/gov/va/med/srcalc/web/view/VariableEntryTest.java [hunk 8]
```diff
@@ -75,28 +102,52 @@ public final void testRetrievedCardiacAge()
...
+    
+    private DiscreteNumericalVariable makeLabVariable(final String key, final ValueRetriever retriever)
+    {
+        final Category wbcWnl = new Category(
+                new MultiSelectOption("WNL"), 11.0f, true);
+        final Category wbcHigh = new Category(
+                new MultiSelectOption(">11.0"), Float.POSITIVE_INFINITY, false);
+        final List<Category> categories = Arrays.asList(wbcWnl, wbcHigh);
+        final DiscreteNumericalVariable var = new DiscreteNumericalVariable(
+                "Generic Lab Display Name", SampleModels.labVariableGroup(), new HashSet<>(categories), key);
```
Comment (dgtombs) [inline srcalc/src/test/java/gov/va/med/srcalc/web/view/VariableEntryTest.java]: Good choice of display name.

**Pull request merged**